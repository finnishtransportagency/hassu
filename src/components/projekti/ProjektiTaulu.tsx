import { VelhoHakuTulos } from "@graphql/apiModel";
import React, { ReactElement } from "react";
import styles from "@styles/ProjektiTaulu.module.css";

interface Props {
  projektit: VelhoHakuTulos[];
  isLoading?: boolean;
}

export default function ProjektiTaulu({ projektit, isLoading }: Props): ReactElement {
  return (
    <table className={styles["project-table"]}>
      <thead>
        <tr>
          <th>Asiatunnus</th>
          <th>Nimi</th>
          <th>Tyyppi</th>
        </tr>
      </thead>
      <tbody>
        {isLoading
          ? Array.from({ length: 3 }, (_, index) => (
              <tr key={index} className="animate-pulse">
                <td className="sm:w-1/4" data-label="Asiatunnus">
                  <div className="bg-gray h-4 w-1/2 md:w-full rounded-md ml-auto mr-0 my-1 align-middle" />
                </td>
                <td className="sm:w-1/2" data-label="Nimi">
                  <div className="bg-gray h-4 w-1/2 md:w-full rounded-md ml-auto mr-0 my-1 align-middle" />
                </td>
                <td className="sm:w-1/4" data-label="Tyyppi">
                  <div className="bg-gray h-4 w-1/2 md:w-full rounded-md ml-auto mr-0 my-1 align-middle" />
                </td>
              </tr>
            ))
          : projektit.map(({ oid, nimi, tyyppi }) => (
              <tr key={oid} className={styles["project-row"]}>
                <td className="sm:w-1/4" data-label="Asiatunnus">
                  {oid}
                </td>
                <td className="sm:w-1/2" data-label="Nimi">
                  {nimi}
                </td>
                <td className="sm:w-1/4" data-label="Tyyppi">
                  {tyyppi}
                </td>
              </tr>
            ))}
      </tbody>
    </table>
  );
}
