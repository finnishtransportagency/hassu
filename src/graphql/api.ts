import { Amplify } from "@aws-amplify/core";

import awsExports from "../aws-exports";
import { API, graphqlOperation } from "aws-amplify";
import { GraphQLAPIClass } from "@aws-amplify/api-graphql";
import * as model from "./apiModel";
import * as queries from "./queries";
import * as mutations from "./mutations";
import log from "loglevel";

import { GraphQLResult } from "@aws-amplify/api-graphql/src/types/index";

Amplify.configure(awsExports);

const yllapitoEndpoint = awsExports.aws_appsync_graphqlEndpoint.replace("/graphql", "/yllapito/graphql");
export const yllapitoAPI = new GraphQLAPIClass({ ...awsExports, aws_appsync_graphqlEndpoint: yllapitoEndpoint });
Amplify.register(yllapitoAPI);
yllapitoAPI.configure({ aws_appsync_graphqlEndpoint: yllapitoEndpoint });

const callYllapitoAPI = async (operation: any) => {
  try {
    return yllapitoAPI.graphql(operation);
  } catch (e) {
    log.error(e);
    window.location.pathname = "/yllapito/kirjaudu";
  }
};

export async function getSuunnitelmaById(suunnitelmaId: string) {
  const response = (await callYllapitoAPI(
    graphqlOperation(queries.getSuunnitelmaById, { suunnitelmaId })
  )) as GraphQLResult<model.GetSuunnitelmaByIdQuery>;
  return response.data?.getSuunnitelmaById as model.Suunnitelma;
}

export async function createSuunnitelma(suunnitelma: model.CreateSuunnitelmaInput) {
  return await callYllapitoAPI(graphqlOperation(mutations.createSuunnitelma, { suunnitelma }));
}

export async function updateSuunnitelma(suunnitelma: model.UpdateSuunnitelmaInput) {
  return await callYllapitoAPI(graphqlOperation(mutations.updateSuunnitelma, { suunnitelma }));
}

export async function getVelhoSuunnitelmasByName(
  suunnitelmaName: string,
  requireExactMatch?: boolean
): Promise<model.VelhoHakuTulos[]> {
  const response = (await callYllapitoAPI(
    graphqlOperation(queries.getVelhoSuunnitelmasByName, { suunnitelmaName, requireExactMatch })
  )) as GraphQLResult<model.GetVelhoSuunnitelmasByNameQuery>;
  return response.data?.getVelhoSuunnitelmasByName as model.VelhoHakuTulos[];
}

export async function listSuunnitelmat() {
  const response = (await API.graphql(
    graphqlOperation(queries.listSuunnitelmat)
  )) as GraphQLResult<model.ListSuunnitelmatQuery>;
  return response.data?.listSuunnitelmat as model.Suunnitelma[];
}

export async function getCurrentUser(): Promise<model.User | undefined> {
  try {
    const response = (await callYllapitoAPI(
      graphqlOperation(queries.getCurrentUser)
    )) as GraphQLResult<model.GetCurrentUserQuery>;
    return response.data?.getCurrentUser as model.User;
  } catch (e) {
    log.debug(e);
  }
}
