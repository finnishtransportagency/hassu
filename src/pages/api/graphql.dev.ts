import type { NextApiRequest, NextApiResponse } from "next";
import { HttpRequest } from "@aws-sdk/protocol-http";
import { HeaderBag } from "@aws-sdk/types";
import { sendSignedRequest } from "../../../backend/src/aws/awsRequest";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  let body = req.body;
  const urlObject = new URL(process.env.APPSYNC_URL || "");
  const isYllapito = req.query.yllapito == "true";

  let headers = {
    host: urlObject.host,
  } as unknown as HeaderBag;
  if (isYllapito) {
    headers["x-hassudev-uid"] = getCookieOrDefault(req.cookies, "x-hassudev-uid", process.env["x-hassudev-uid"]) as string;
    headers["x-hassudev-roles"] = getCookieOrDefault(req.cookies, "x-hassudev-roles", process.env["x-hassudev-roles"]) as string;
  }
  const request = new HttpRequest({
    headers: headers,
    hostname: urlObject.hostname,
    method: "POST",
    path: urlObject.pathname,
    body: JSON.stringify(body),
  });
  const { body: responseBody, statusCode } = await sendSignedRequest(request, "appsync");

  await res.status(statusCode).json(responseBody);
}

function getCookieOrDefault(cookies: Partial<{ [p: string]: string }>, name: string, defaultValue: string | undefined) {
  let value = cookies?.[name];
  if (value) {
    return value;
  }
  return defaultValue;
}
