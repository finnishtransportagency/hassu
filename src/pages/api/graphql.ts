import type { NextApiRequest, NextApiResponse } from "next";
import { HttpRequest } from "@aws-sdk/protocol-http";
import { HeaderBag } from "@aws-sdk/types";
import { sendSignedRequest } from "../../../backend/src/aws/awsRequest";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  let body = req.body;
  const urlObject = new URL(process.env.APPSYNC_URL || "");
  const isYllapito = req.query.yllapito == "true";

  let params: URLSearchParams | undefined;
  if (typeof window !== "undefined") {
    params = window.location?.search ? new URLSearchParams(window.location.search) : undefined;
  }
  let headers = {
    host: urlObject.host,
  } as unknown as HeaderBag;
  if (isYllapito) {
    headers["x-hassudev-uid"] = getParamOrDefault(params, "x-hassudev-uid", process.env["x-hassudev-uid"]) as string;
    headers["x-hassudev-roles"] = getParamOrDefault(
      params,
      "x-hassudev-roles",
      process.env["x-hassudev-roles"]
    ) as string;
  }
  const request = new HttpRequest({
    headers: headers,
    hostname: urlObject.hostname,
    method: "POST",
    path: urlObject.pathname,
    body: JSON.stringify(body),
  });
  const { body: responseBody, statusCode } = await sendSignedRequest(request, "appsync");

  res.status(statusCode).json(JSON.stringify(responseBody));
}

function storeValue(name: string, value: string) {
  if (typeof window !== "undefined") {
    localStorage.setItem(name, value);
  }
}

function getValue(name: string) {
  if (typeof window !== "undefined") {
    return localStorage.getItem(name);
  }
}

function getParamOrDefault(params: URLSearchParams | undefined, name: string, defaultValue: string | undefined) {
  if (params) {
    if (params.has(name)) {
      const value = params.get(name) || "";
      storeValue(name, value);
      return value;
    }
  }
  const valueFromStorage = getValue(name);
  if (!!valueFromStorage) {
    return valueFromStorage;
  }
  return defaultValue;
}
