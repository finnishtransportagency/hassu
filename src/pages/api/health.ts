import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  res.status(200).json({ status: "ok" });
}
