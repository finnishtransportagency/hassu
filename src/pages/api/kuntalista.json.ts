import type { NextApiRequest, NextApiResponse } from "next";
import { S3Cache } from "../../../backend/src/cache/s3Cache";
import log from "loglevel";
import { setupLambdaMonitoring, wrapXrayAsync } from "../../../backend/src/aws/monitoring";

const KUNTALISTA_TTL_SECONDS = 24 * 3 * 60 * 60;

async function fetchKuntaLista() {
  const response = await fetch("http://rajapinnat.ymparisto.fi/api/Hakemistorajapinta/1.0/odata/Kunta");

  const data: any = await response.json();
  const list = data?.value?.map((kunta: any) => {
    return { label: kunta.Nimi, value: kunta.Nimi.toUpperCase() };
  });
  list.unshift({ label: "", value: "" });
  return list;
}

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  setupLambdaMonitoring();
  return await wrapXrayAsync("handler", async () => {
    const s3Cache = new S3Cache(KUNTALISTA_TTL_SECONDS);
    const kuntaList: Record<string, string> = await s3Cache.get(
      "kuntalista",
      KUNTALISTA_TTL_SECONDS * 1000,
      async () => {
        log.info("Updating kuntalista, it has been expired");
        const list = await fetchKuntaLista();
        s3Cache.put("kuntalista", list);
      },
      async () => {
        log.info("Updating kuntalista, it is missing");
        const list = await fetchKuntaLista();
        s3Cache.put("kuntalista", list);
        return list;
      }
    );

    res.setHeader(
      "Cache-Control",
      "public, s-maxage=" + KUNTALISTA_TTL_SECONDS + ", stale-while-revalidate=" + (KUNTALISTA_TTL_SECONDS - 30)
    );
    res.status(200).json(kuntaList);
  });
}
