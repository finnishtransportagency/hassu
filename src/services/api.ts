import getConfig from "next/config";
import { AbstractApi } from "../../common/abstractApi";

const { publicRuntimeConfig } = getConfig();
const apiImpl = publicRuntimeConfig.apiImpl;
export const api: AbstractApi = require("./api/" + apiImpl).api;

export { apiConfig } from "../../common/abstractApi";
export * from "../../common/graphql/apiModel";
