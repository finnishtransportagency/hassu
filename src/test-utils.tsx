import React, { ReactNode } from "react";
import I18nProvider from "next-translate/I18nProvider";
import i18nConfig, { defaultLocale } from "../i18n";
import loadNamespaces from "next-translate/loadNamespaces";
import { I18nDictionary } from "next-translate";

const getAllNamespaces = () =>
  Object.values(i18nConfig.pages).reduce((namespaces, pageNamespaces) => {
    pageNamespaces.forEach((item) => !namespaces.some((ns) => ns === item) && namespaces.push(item));
    return namespaces;
  }, []);

export async function componentWithTranslation(children: ReactNode) {
  const config = {
    ...i18nConfig,
    pathname: "*",
    pages: {
      "*": getAllNamespaces(),
    },
    loader: false,
  };
  const { __namespaces } = await loadNamespaces(config);
  return (
    <I18nProvider lang={defaultLocale} namespaces={__namespaces as Record<string, I18nDictionary>}>
      {children}
    </I18nProvider>
  );
}
