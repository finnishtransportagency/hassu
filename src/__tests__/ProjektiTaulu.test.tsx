/**
 * @jest-environment jsdom
 */

import React from "react";
import ProjektiTaulu from "@components/projekti/ProjektiTaulu";
import { ProjektiTyyppi, VelhoHakuTulos } from "@services/api";
import renderer from "react-test-renderer";
import { componentWithTranslation } from "../test-utils";

describe("ProjektiTaulu", () => {
  const props = {
    isLoading: false,
    projektit: [
      {
        __typename: "VelhoHakuTulos",
        oid: "1234",
        tyyppi: "TIE" as ProjektiTyyppi.TIE,
        nimi: "Tampereen Tie Projekti",
      },
    ] as VelhoHakuTulos[],
  };

  it("renders ProjektiTaulu unchanged", async () => {
    const tree = renderer.create(await componentWithTranslation(<ProjektiTaulu {...props} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
