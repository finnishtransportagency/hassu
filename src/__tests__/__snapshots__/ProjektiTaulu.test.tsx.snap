// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjektiTaulu renders ProjektiTaulu unchanged 1`] = `
<table
  className="project-table"
>
  <thead>
    <tr>
      <th>
        Tunnus
      </th>
      <th>
        Nimi
      </th>
      <th>
        Tyyppi
      </th>
      <th>
        Projektipäällikkö
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="project-row"
    >
      <td
        className="sm:w-2/12"
        data-label="Tunnus"
      >
        1234
      </td>
      <td
        className="sm:w-6/12"
        data-label="Nimi"
      >
        Tampereen Tie Projekti
      </td>
      <td
        className="sm:w-2/12"
        data-label="Tyyppi"
      >
        Tiesuunnitelma
      </td>
      <td
        className="sm:w-2/12"
        data-label="Projektipäällikkö"
      >
        -
      </td>
    </tr>
  </tbody>
</table>
`;
